Popup enrichment from wikidata

jsanz opened the pull request:
This is a prototype since I don't have much knowledge about the display but it should be a good starting point if it's accepted.

![Peek 2020-11-06 10-23](https://user-images.githubusercontent.com/188264/98349282-1811f280-201a-11eb-9652-7ff0b0530d41.gif)

Closes #30

jsanz commented:
@jessisena let me know what you think

jessisena commented on src/services/GeojsonService.js:
```diff
@@ -25,13 +27,49 @@ export default class GeojsonMapService {
-  static getHTMLWikipediaLink(link, popupText) {
-    if (link !== '') {
-      return `<p class=""><a  class="btn btn-light" target="_blank" href='${link}'><i class="fab fa-wikipedia-w"></i></a></p>`;
+  static getHTMLWikipediaLink(link, popupText, data) {
```
WDYT about renaming `data` to something a bit more auto explicative like `wikidataDetails`, `dataDetails`, or similar?

jessisena commented on src/services/GeojsonService.js:
```diff
@@ -110,20 +148,31 @@ export default class GeojsonMapService {
-        const html = `<div class="row"><div class="col-sm"><div class="${popupType}"><p>${name}</p>
-          ${gender === FEMALE ? this.getHTMLWikipediaLink(link, popupText) : ''}</div></div></div>`;
+        // eslint-disable-next-line no-console
+        getWikdataDetails(link).then((data) => {
```
Same as before, WDYT about re-naming this `data` to something a bit more explicative? 

And just to be sure, `getWikdataDetails` is a typo, or did you mean `getWikidataDetails`?

jessisena commented on src/services/WikidataService.js:
```diff
@@ -0,0 +1,35 @@
+/* eslint-disable no-console */
+
+// eslint-disable-next-line max-len
+const sparqlQueryTemplate = `SELECT%20%0A%20%20%3Fid%20%0A%20%20(%3FidLabel%20AS%20%3Fname)%20%0A%20%20(%3Fdesc%20as%20%3Fdescription)%20%0A%20%20(%3FgenderLabel%20AS%20%3Fgender)%0A%20%20(SAMPLE(%3Fbirths)%20AS%20%3Fbirth)%20%0A%20%20(SAMPLE(%3Fdeaths)%20AS%20%3Fdeath)%20%0A%20%20(SAMPLE(%3Fpic)%20AS%20%3Fpicture)%20%0A%20%20(GROUP_CONCAT(DISTINCT%20%3FoccupationsLabel%3B%20SEPARATOR%20%3D%20%22 ...
```
Would it make sense to add this one in the `Constants.js` file? 🙄

jessisena commented on src/services/WikidataService.js:
```diff
@@ -0,0 +1,35 @@
+  const { length } = wikipediaLink;
+  const nameLink = wikipediaLink.substring(wikipediaLink.lastIndexOf('/') + 1, length).replace(/_/g, '%20');
+  const query = sparqlQueryTemplate.replace('##NAME##', nameLink);
+  const url = `https://query.wikidata.org/sparql?format=json&query=${query}`;
```
Same here, would it  make sense adding this URL in the `Constants.js` file? 

`const wikidateSparwlUrl = 'https://query.wikidata.org/sparql?format=json&query='`

jessisena commented on src/services/WikidataService.js:
```diff
@@ -0,0 +1,35 @@
+  return new Promise((resolve, reject) => {
+    fetch(url)
+      .then((response) => response.json())
+      .then((data) => {
+        if ('results' in data && 'bindings' in data.results) {
```
I'm thinking about simplifying this logic a bit to something like:

```
.then(({results}) => {
if (results?.bindings) {
```

jsanz replied:
done, but I had to add a babel plugin and a new `.babelrc` file to specify it.

jessisena reviewed (commented):
@jsanz this is amazing! 👏 thanks!  I've let some little comments in the code, but as you said I think is a great starter point.

I have only 2 concerns about the popup:
- the "flashing effect" while loading plus the popup changing size, and
- the best way to visualize the new data in small devices/screens

For the 1st one, I'm thinking that maybe is worth it to fill the popup from the beginning with some kind of placeholders where the final data should appear. And, in case eventually the data is not there, it's also a way to show this fact (the same way that the Wikipedia button is being shown disabled when there's no wikipedia link available).

Other ideas, opinions and/or suggestions are welcome 🙂

jessisena commented:
@jsanz this is a request, but could you include a brief explanation (similar to the one in #30 ) about the sparql query inside our README? 🙏

jsanz commented:
Thanks for the review @jessisena, I understand then this functionality is desirable.

* I'll definitely address your comments and study more the rest of the code to try to adapt to current practices.
* Sure, I will add a detailed explanation of the SPARQL query to the README as part of this PR. My idea is that this query is an intermediate solution since we could switch many of the current Wikipedia links in the generated data by Wikidata identifiers but that's a different task.

Regarding the UI:

* I will also handle the mobile use case
* I will work on a popup that has all the placeholders by default, and when the user clicks try to get filled, showing some notice about missing data. Notice that sometimes is OK to have missing data, like birth/death dates for goddesses, for example. I like the idea of also advocating on missing data in Wikidata directly on the site.
* I was thinking about maybe using the picture as a background for the full pop up, I will follow up with some proposals if I need your advice on different options.

Anything else I should take care of?

jessisena commented:
Yes, this feature is absolutely desirable @jsanz 🚀  Thanks for addressing the comments and your willingness to add the SPARQL explanation. 

And about the UI, I've just reviewed the popup (mobile and desktop) and actually I think we could increase and establish the same size for both without problems. For example, the width could be something between 240px or 200px, WDYT?

And about the design, my concern about adding the image as a background for the full popup is readability and visibility in general 🙄 but you can tell me better once you try it 👍 

Actually, I was thinking about a full re-style of the popup to something like this horizontal design:
![image](https://user-images.githubusercontent.com/3327279/98869498-c8905400-2469-11eb-8ac4-26fef3df0e28.png)

or another option is to keep the verticality (common style):
![image](https://user-images.githubusercontent.com/3327279/98869762-33418f80-246a-11eb-85db-699c537d27ee.png)


Although I personally prefer the 1st one 🙂

jsanz commented:
I'll give the image as a background a try and if I see it readable I'll share it.

I like the _common style_ option because we can use it both for mobile and desktop but it shouldn't hard to implement a responsive layout.

jsanz commented:
@jessisena please take a look at the changes:

* Added a caching system to avoid repeated requests to wikidata
* Introduced [Handlebars](https://handlebarsjs.com/) to manage the two templates used
* Pop ups are "enriched" on hover thanks to :point_up: 
* Bigger layout following your suggestion of having the picture on the left

It's not perfect and there are some cases still to handle like a better signaling when no data has been retrieved from wikidata, but I think it's ready for you to check

![Peek 2020-11-20 20-01](https://user-images.githubusercontent.com/188264/99839746-8bb01580-2b6b-11eb-9f77-5e97993f44d6.gif)

jsanz commented:
@jessisena last commit ab6f56b added:

* Increased the popup height to avoid the small jump (it still may happen to extremely narrow images)
* Treats the case for when instead of `wikipedia_link` there is a `wikidata` field
* Splits the previous cache into two:

  1. One that stores the wikidata IDs corresponding to women names
  2. Another that stores the wikidata details based on wikidata IDs

  This way the pop up enrichment can be a one step if the geojson stores the wikidata ID or a two steps process when the input is a Wikipedia name: first it retrieves the wikidata ID and then it retrieves the rest of the details. This way we have a two level cache to deal with both scenarios and saving as many roundtrips to wikidata as possible.

The queries to Wikidata now are not encoded and easier to read. I still need to document them somewhere. I'll try to do this ASAP but the rest of the code is ready for proper review.

jsanz commented:
BTW the caches are now for a single day, we may want to extend that a bit but totally up to you. At the end of the `Constants.js` file: 

```js
export const wikidataExpireCache = 1 * 24 * 3600 * 1000;
```

jessisena commented on src/styles.css:
```diff
@@ -407,33 +407,82 @@ color: #FFFF00;
-  width: 30px;
+
+.popup-female .picture img{
+  width: 100%;
```
I suggest here the use of object-fit and the definition a heigth, even if then image is not fully visualized. This way we keep the ratio across the diferrent popups. 

```suggestion
  width: 100%;
  height: 250px;
  object-fit: cover;
```

jessisena commented on src/styles.css:
```diff
@@ -407,33 +407,82 @@ color: #FFFF00;
+}
+
+.popup-female .picture{
+  min-height: 250px;
```
Because of the previous suggestion, I would delete this one here
```suggestion
```

jsanz commented:
Much better @jessisena, thanks for the suggestions. I've also added details about the queries in the README files. Let me know if there's anything else to improve.

About the Wikidata cache, is **one day** enough, or should we increase it? My reasoning for that figure is that the website is mostly visited by occasional users so saving repeated queries on a single session seems fine to me, but maybe you think it's better to increase the data retention :thinking:

jsanz commented:
@jessisena @fpsampayo I've enabled the deploy preview for this PR, and it is available here

https://deploy-preview-31--geochicasosm-lascallesdelasmujeres.netlify.app/

What do you think about merging this change?

fpsampayo commented:
Perfect for me. Wait only to merge the #82

jsanz commented:
@fpsampayo this should be ready for review, merged with `master` after #82 deploy preview regenerated

https://deploy-preview-31--geochicasosm-lascallesdelasmujeres.netlify.app/

@clds1 @mapanauta you may also want to take a look at the preview of the website with the improved popups.
